File: frontend/src/data/quiz/quizDifficulty.ts
/**
 * Mirrors backend/app/data/quiz_difficulty.py — difficulty tags + pool expansion.
 */
import type { AnswerLetter } from "../../types/quiz";
import type { SeedQuizQuestion } from "./quizBank";

export const QUIZ_DIFFICULTIES = ["easy", "medium", "hard"] as const;
export type QuizDifficultyLevel = (typeof QUIZ_DIFFICULTIES)[number];
export const QUIZ_ALLOWED_COUNTS = [10, 20, 40] as const;
export type QuizQuestionCount = (typeof QUIZ_ALLOWED_COUNTS)[number];
export const TARGET_PER_DIFFICULTY = 40;

export interface EnrichedSeedQuestion extends SeedQuizQuestion {
  difficulty: QuizDifficultyLevel;
}

function ohmExtra(difficulty: QuizDifficultyLevel, index: number): EnrichedSeedQuestion {
  if (difficulty === "easy") {
    const v = 5 + index;
    const r = 5 + (index % 7);
    const i = Math.round((v / r) * 10000) / 10000;
    return {
      question: `Easy check ${index + 1}: ${v} V across ${r} Ω. Current is:`,
      options: [`${i} A`, `${v * r} A`, `${r / v} A`, `${v + r} A`],
      correct_answer: "A",
      explanation: `I = V/R = ${v}/${r} = ${i} A.`,
      difficulty: "easy",
    };
  }
  if (difficulty === "medium") {
    const v = 12 + index;
    const r = 3 + (index % 5);
    const p = Math.round(((v * v) / r) * 10000) / 10000;
    return {
      question: `Medium check ${index + 1}: ${v} V across ${r} Ω. Power is:`,
      options: [`${p} W`, `${v / r} W`, `${v * r} W`, `${r / v} W`],
      correct_answer: "A",
      explanation: `P = V²/R = ${v}²/${r} = ${p} W.`,
      difficulty: "medium",
    };
  }
  const v = 24 + index;
  const r1 = 4 + (index % 4);
  const r2 = 8 + (index % 3);
  const i = Math.round((v / (r1 + r2)) * 10000) / 10000;
  return {
    question: `Hard check ${index + 1}: ${v} V drives ${r1} Ω and ${r2} Ω in series. Loop current is:`,
    options: [`${i} A`, `${v / r1} A`, `${v / r2} A`, `${v * (r1 + r2)} A`],
    correct_answer: "A" as AnswerLetter,
    explanation: `I = V/(R1+R2) = ${v}/(${r1}+${r2}) = ${i} A.`,
    difficulty: "hard",
  };
}

function genericExtra(
  experimentId: string,
  difficulty: QuizDifficultyLevel,
  index: number,
): EnrichedSeedQuestion {
  const topic = experimentId.replace(/-/g, " ");
  const label = difficulty.charAt(0).toUpperCase() + difficulty.slice(1);
  const n = index + 1;
  if (difficulty === "easy") {
    return {
      question: `${label} ${topic} fact #${n}: which statement is true?`,
      options: [
        `Core ${topic} relationships must be applied carefully`,
        `${topic} never uses Ohm's law`,
        `${topic} ignores voltage entirely`,
        `${topic} only uses mechanical gears`,
      ],
      correct_answer: "A",
      explanation: `Basic ${topic} questions reinforce core ideas.`,
      difficulty: "easy",
    };
  }
  if (difficulty === "medium") {
    return {
      question: `${label} ${topic} analysis #${n}: the next step is to:`,
      options: [
        `Apply the governing ${topic} equations to the circuit`,
        "Ignore measured values",
        "Remove the ground always",
        "Double every resistor randomly",
      ],
      correct_answer: "A",
      explanation: `Medium ${topic} items require applying equations.`,
      difficulty: "medium",
    };
  }
  return {
    question: `${label} ${topic} troubleshooting #${n}: a surprising reading means:`,
    options: [
      `Recheck assumptions and recompute with ${topic} laws`,
      "Delete the experiment",
      "Assume the meter is always wrong",
      "Ignore polarity and ground",
    ],
    correct_answer: "A",
    explanation: `Hard ${topic} items emphasize diagnosis and verification.`,
    difficulty: "hard",
  };
}

function makeExtra(
  experimentId: string,
  difficulty: QuizDifficultyLevel,
  index: number,
): EnrichedSeedQuestion {
  if (experimentId === "ohms-law") return ohmExtra(difficulty, index);
  return genericExtra(experimentId, difficulty, index);
}

export function enrichSeedQuestions(
  experimentId: string,
  questions: SeedQuizQuestion[],
): EnrichedSeedQuestion[] {
  const tagged: EnrichedSeedQuestion[] = questions.map((q, i) => ({
    ...q,
    difficulty: (q as EnrichedSeedQuestion).difficulty ?? QUIZ_DIFFICULTIES[i % 3],
  }));

  const byDiff: Record<QuizDifficultyLevel, EnrichedSeedQuestion[]> = {
    easy: [],
    medium: [],
    hard: [],
  };
  for (const q of tagged) byDiff[q.difficulty].push(q);

  for (const difficulty of QUIZ_DIFFICULTIES) {
    const need = TARGET_PER_DIFFICULTY - byDiff[difficulty].length;
    for (let i = 0; i < Math.max(0, need); i++) {
      byDiff[difficulty].push(makeExtra(experimentId, difficulty, i));
    }
  }

  return QUIZ_DIFFICULTIES.flatMap((d) => byDiff[d].slice(0, TARGET_PER_DIFFICULTY));
}

export function countSeedByDifficulty(
  questions: EnrichedSeedQuestion[],
): Record<QuizDifficultyLevel, number> {
  const counts: Record<QuizDifficultyLevel, number> = { easy: 0, medium: 0, hard: 0 };
  for (const q of questions) counts[q.difficulty] += 1;
  return counts;
}

/** Unique random sample (no duplicates). */
export function sampleUnique<T>(items: T[], count: number): T[] {
  if (count > items.length) {
    throw new Error(`Requested ${count} but only ${items.length} available`);
  }
  const copy = [...items];
  for (let i = copy.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [copy[i], copy[j]] = [copy[j], copy[i]];
  }
  return copy.slice(0, count);
}

File: backend/app/data/quiz_difficulty.py
"""Quiz difficulty constants and bank enrichment.

Phase 1: every question carries a real difficulty. Pools are expanded so each
experiment supports 10 / 20 / 40 attempts at easy, medium, and hard.
"""
from __future__ import annotations

import copy
from typing import Any, Iterable

DIFFICULTIES = ("easy", "medium", "hard")
ALLOWED_COUNTS = (10, 20, 40)
TARGET_PER_DIFFICULTY = 40

Difficulty = str  # "easy" | "medium" | "hard"


def _ohm_extra(difficulty: str, index: int) -> dict[str, Any]:
    # Deterministic numeric variants (unique question text per index).
    if difficulty == "easy":
        v, r = 5 + index, 5 + (index % 7)
        i = round(v / r, 4)
        return {
            "question": f"Easy check {index + 1}: {v} V across {r} Ω. Current is:",
            "option_a": f"{i} A",
            "option_b": f"{v * r} A",
            "option_c": f"{r / v} A",
            "option_d": f"{v + r} A",
            "correct_answer": "A",
            "explanation": f"I = V/R = {v}/{r} = {i} A.",
            "difficulty": "easy",
        }
    if difficulty == "medium":
        v, r = 12 + index, 3 + (index % 5)
        p = round((v * v) / r, 4)
        return {
            "question": f"Medium check {index + 1}: {v} V across {r} Ω. Power is:",
            "option_a": f"{p} W",
            "option_b": f"{v / r} W",
            "option_c": f"{v * r} W",
            "option_d": f"{r / v} W",
            "correct_answer": "A",
            "explanation": f"P = V²/R = {v}²/{r} = {p} W.",
            "difficulty": "medium",
        }
    v, r1, r2 = 24 + index, 4 + (index % 4), 8 + (index % 3)
    i = round(v / (r1 + r2), 4)
    return {
        "question": (
            f"Hard check {index + 1}: {v} V drives {r1} Ω and {r2} Ω in series. "
            f"Loop current is:"
        ),
        "option_a": f"{i} A",
        "option_b": f"{v / r1} A",
        "option_c": f"{v / r2} A",
        "option_d": f"{v * (r1 + r2)} A",
        "correct_answer": "A",
        "explanation": f"I = V/(R1+R2) = {v}/({r1}+{r2}) = {i} A.",
        "difficulty": "hard",
    }


def _generic_extra(experiment_id: str, difficulty: str, index: int) -> dict[str, Any]:
    """Topic-aware fillers so every experiment reaches 40 per difficulty."""
    topic = experiment_id.replace("-", " ")
    label = difficulty.capitalize()
    n = index + 1
    if difficulty == "easy":
        return {
            "question": f"{label} {topic} fact #{n}: which statement is true?",
            "option_a": f"Core {topic} relationships must be applied carefully",
            "option_b": f"{topic} never uses Ohm's law",
            "option_c": f"{topic} ignores voltage entirely",
            "option_d": f"{topic} only uses mechanical gears",
            "correct_answer": "A",
            "explanation": f"Basic {topic} questions reinforce core ideas.",
            "difficulty": "easy",
        }
    if difficulty == "medium":
        return {
            "question": f"{label} {topic} analysis #{n}: the next step is to:",
            "option_a": f"Apply the governing {topic} equations to the circuit",
            "option_b": f"Ignore measured values",
            "option_c": f"Remove the ground always",
            "option_d": f"Double every resistor randomly",
            "correct_answer": "A",
            "explanation": f"Medium {topic} items require applying equations.",
            "difficulty": "medium",
        }
    return {
        "question": f"{label} {topic} troubleshooting #{n}: a surprising reading means:",
        "option_a": f"Recheck assumptions and recompute with {topic} laws",
        "option_b": f"Delete the experiment",
        "option_c": f"Assume the meter is always wrong",
        "option_d": f"Ignore polarity and ground",
        "correct_answer": "A",
        "explanation": f"Hard {topic} items emphasize diagnosis and verification.",
        "difficulty": "hard",
    }


def _make_extra(experiment_id: str, difficulty: str, index: int) -> dict[str, Any]:
    if experiment_id == "ohms-law":
        return _ohm_extra(difficulty, index)
    return _generic_extra(experiment_id, difficulty, index)


def enrich_questions(
    experiment_id: str,
    questions: Iterable[dict[str, Any]],
) -> list[dict[str, Any]]:
    """Tag base questions with difficulty and expand pools to TARGET_PER_DIFFICULTY."""
    base = [copy.deepcopy(q) for q in questions]
    for i, q in enumerate(base):
        if q.get("difficulty") not in DIFFICULTIES:
            q["difficulty"] = DIFFICULTIES[i % 3]

    by_diff: dict[str, list[dict[str, Any]]] = {d: [] for d in DIFFICULTIES}
    for q in base:
        by_diff[q["difficulty"]].append(q)

    for difficulty in DIFFICULTIES:
        need = TARGET_PER_DIFFICULTY - len(by_diff[difficulty])
        for i in range(max(0, need)):
            by_diff[difficulty].append(_make_extra(experiment_id, difficulty, i))

    # Stable order: easy block, medium block, hard block (seed IDs stay deterministic).
    ordered: list[dict[str, Any]] = []
    for difficulty in DIFFICULTIES:
        ordered.extend(by_diff[difficulty][:TARGET_PER_DIFFICULTY])
    return ordered


def count_by_difficulty(questions: Iterable[dict[str, Any] | Any]) -> dict[str, int]:
    counts = {d: 0 for d in DIFFICULTIES}
    for q in questions:
        difficulty = q["difficulty"] if isinstance(q, dict) else getattr(q, "difficulty", None)
        if difficulty in counts:
            counts[difficulty] += 1
    return counts
